`TestDifferentialEvolutionSolver` failures in CI

The `differential_evolution` tests are probably the most flaky tests we have, so I thought it's time to open a dedicated issue for it after seeing the same failure in the `scipy-release` CI a couple of times.

`TestDifferentialEvolutionSolver.test_L4` is failing on macOS x86-64, e.g. from [this `scipy-release` log](https://github.com/scipy/scipy-release/actions/runs/21819557437/job/62949109307):

```
  ___________________ TestDifferentialEvolutionSolver.test_L4 ____________________
  ../venv-test-x86_64/lib/python3.14/site-packages/scipy/optimize/tests/test__differential_evolution.py:1246: in test_L4
      assert_allclose(res.fun, f_opt, atol=0.001)
  E   AssertionError: 
  E   Not equal to tolerance rtol=1e-07, atol=0.001
  E   
  E   Mismatched elements: 1 / 1 (100%)
  E   Max absolute difference among violations: 1049.80977247
  E   Max relative difference among violations: 0.14892507
  E    ACTUAL: array(8099.057772)
  E    DESIRED: array(7049.248)
          A          = array([[ 0.    ,  0.    ,  0.    ,  0.0025,  0.    ,  0.0025,  0.    ,
           0.    ],
         [ 0.    ,  0.    ,  0.... ,  0.0025,
           0.    ],
         [ 0.    ,  0.    ,  0.    ,  0.    , -0.01  ,  0.    ,  0.    ,
           0.01  ]])
          L          = <scipy.optimize._constraints.LinearConstraint object at 0x113c86a50>
          N          = <scipy.optimize._constraints.NonlinearConstraint object at 0x123193ec0>
          b          = array([1, 1, 1])
          bounds     = [(100, 10000), (1000, 10000), (1000, 10000), (10, 1000), (10, 1000), (10, 1000), ...]
          c1         = <function TestDifferentialEvolutionSolver.test_L4.<locals>.c1 at 0x1222e1d20>
          constraints = (<scipy.optimize._constraints.LinearConstraint object at 0x113c86a50>, <scipy.optimize._constraints.NonlinearConstraint object at 0x123193ec0>)
          f          = <function TestDifferentialEvolutionSolver.test_L4.<locals>.f at 0x1222e26c0>
          f_opt      = 7049.248
          res        =              message: Optimization terminated successfully.
               success: True
                   fun: 8099.0577...000e+00,  0.000e+00]), array([ 0.000e+00,  0.000e+00,  0.000e+00])]
      constr_violation: 0.0
                 maxcv: 0.0
          self       = <scipy.optimize.tests.test__differential_evolution.TestDifferentialEvolutionSolver object at 0x113c87e70>
          x_opt      = [579.306692, 1359.97063, 5109.9707, 182.0177, 295.601172, 217.9823, ...]
```

That same `test_L4` on Guix, as discussed in gh-22595. There @andyfaff said: 

> The `optimize/tests/test__differential_evolution.py::TestDifferentialEvolutionSolver::test_L4` fail is because solver options have been set such that minimal calculation is used. These tight settings have caused the solver to fail sometimes. Higher popsize will result in a test that will pass, but will take  longer to run.

I'm not sure that this can be the whole story, the test is supposed to be deterministic (at first sight the `rng` plumbing looks fine); small sample sizes shouldn't normally cause such large differences, unless the test is numerically unstable. 

* * *

There are two `differential_evolution` tests mentioned in gh-22758 that are flaky. Inspecting those tests, it is clear that the function is called without any seeding. And actually many more tests in `test__differential_evolution.py` are failing to pass in `rng` (or `seed`). This seems to be the root of the broader problem. As a rule, for all tests that call functionality that draws random numbers, seeding must be used for test suite stability. So we should go through that file and fix this up.

--- a/scipy/optimize/tests/test__differential_evolution.py
+++ b/scipy/optimize/tests/test__differential_evolution.py
@@ -1205,6 +1205,13 @@ class TestDifferentialEvolutionSolver:
         assert_(np.all(res.x >= np.array(bounds)[:, 0]))
         assert_(np.all(res.x <= np.array(bounds)[:, 1]))
 
+    """
+    # test commented out because of test flakes in macosx-x86_64 that
+    # were also seen in GUIX. It's suspected that they're due to
+    # numeric instability. Given that constraint handling is
+    # adequately tested in the other problems we can skip this one.
+    # It's left in the test suite in case we want to use it again
+    # in the future (and a reminder what the problems with it were).
     @pytest.mark.fail_slow(10)
     def test_L4(self):
         # Lampinen ([5]) test problem 4
@@ -1257,6 +1264,7 @@ class TestDifferentialEvolutionSolver:
         assert_(np.all(np.array(c1(res.x)) >= 0))
         assert_(np.all(res.x >= np.array(bounds)[:, 0]))
         assert_(np.all(res.x <= np.array(bounds)[:, 1]))
+    """
 
     @pytest.mark.fail_slow(10)
     def test_L5(self):

--- a/scipy/optimize/tests/test__differential_evolution.py
+++ b/scipy/optimize/tests/test__differential_evolution.py
@@ -1041,6 +1041,7 @@ class TestDifferentialEvolutionSolver:
     @pytest.mark.fail_slow(20)
     def test_L1(self):
         # Lampinen ([5]) test problem 1
+        rng = np.random.default_rng(12345)
 
         def f(x):
             x = np.hstack(([0], x))  # 1-indexed to match reference
@@ -1067,7 +1068,7 @@ class TestDifferentialEvolutionSolver:
 
         # using a lower popsize to speed the test up
         res = differential_evolution(
-            f, bounds, strategy='best1bin', rng=12345, constraints=(L,),
+            f, bounds, strategy='best1bin', rng=rng, constraints=(L,),
             popsize=5, tol=0.01
         )
 
@@ -1089,8 +1090,9 @@ class TestDifferentialEvolutionSolver:
         L = LinearConstraint(csr_array(A), -np.inf, b)
 
         # using a lower popsize to speed the test up
+        rng = np.random.default_rng(12345)
         res = differential_evolution(
-            f, bounds, strategy='best1bin', rng=1211134, constraints=(L,),
+            f, bounds, strategy='best1bin', rng=rng, constraints=(L,),
             popsize=2, tol=0.05
         )
 
@@ -1121,9 +1123,10 @@ class TestDifferentialEvolutionSolver:
         constraints = (L, N, L2, N2)
 
         with warnings.catch_warnings():
+            rng = np.random.default_rng(12345)
             warnings.simplefilter("ignore", UserWarning)
             res = differential_evolution(
-                f, bounds, strategy='best1bin', rng=1211134,
+                f, bounds, strategy='best1bin', rng=rng,
                 constraints=constraints, popsize=2, tol=0.05
             )
 
@@ -1136,6 +1139,7 @@ class TestDifferentialEvolutionSolver:
     @pytest.mark.fail_slow(10)
     def test_L2(self):
         # Lampinen ([5]) test problem 2
+        rng = np.random.default_rng(111121332)
 
         def f(x):
             x = np.hstack(([0], x))  # 1-indexed to match reference
@@ -1159,7 +1163,7 @@ class TestDifferentialEvolutionSolver:
         with warnings.catch_warnings():
             warnings.simplefilter("ignore", UserWarning)
             res = differential_evolution(f, bounds, strategy='best1bin',
-                                         rng=1234, constraints=constraints)
+                                         rng=rng, constraints=constraints)
 
         f_opt = 680.6300599487869
         x_opt = (2.330499, 1.951372, -0.4775414, 4.365726,
@@ -1176,6 +1180,7 @@ class TestDifferentialEvolutionSolver:
     @pytest.mark.fail_slow(10)
     def test_L3(self):
         # Lampinen ([5]) test problem 3
+        rng = np.random.default_rng(1234520312198)
 
         def f(x):
             x = np.hstack(([0], x))  # 1-indexed to match reference
@@ -1208,7 +1213,7 @@ class TestDifferentialEvolutionSolver:
 
         with warnings.catch_warnings():
             warnings.simplefilter("ignore", UserWarning)
-            res = differential_evolution(f, bounds, rng=1234,
+            res = differential_evolution(f, bounds, rng=rng,
                                          constraints=constraints, popsize=3)
 
         x_opt = (2.171996, 2.363683, 8.773926, 5.095984, 0.9906548,
@@ -1288,7 +1293,7 @@ class TestDifferentialEvolutionSolver:
     @pytest.mark.fail_slow(10)
     def test_L5(self):
         # Lampinen ([5]) test problem 5
-
+        rng = np.random.default_rng(209331234)
         def f(x):
             x = np.hstack(([0], x))  # 1-indexed to match reference
             fun = (np.sin(2*np.pi*x[1])**3*np.sin(2*np.pi*x[2]) /
@@ -1304,7 +1309,7 @@ class TestDifferentialEvolutionSolver:
         bounds = [(0, 10)]*2
         constraints = (N)
 
-        res = differential_evolution(f, bounds, strategy='rand1bin', rng=1234,
+        res = differential_evolution(f, bounds, strategy='rand1bin', rng=rng,
                                      constraints=constraints)
 
         x_opt = (1.22797135, 4.24537337)
@@ -1319,6 +1324,8 @@ class TestDifferentialEvolutionSolver:
     @pytest.mark.fail_slow(10)
     def test_L6(self):
         # Lampinen ([5]) test problem 6
+        rng = np.random.default_rng(209331234)
+
         def f(x):
             x = np.hstack(([0], x))  # 1-indexed to match reference
             fun = (x[1]-10)**3 + (x[2] - 20)**3
@@ -1332,7 +1339,7 @@ class TestDifferentialEvolutionSolver:
         N = NonlinearConstraint(c1, 0, np.inf)
         bounds = [(13, 100), (0, 100)]
         constraints = (N)
-        res = differential_evolution(f, bounds, strategy='rand1bin', rng=1234,
+        res = differential_evolution(f, bounds, strategy='rand1bin', rng=rng,
                                      constraints=constraints, tol=1e-7)
         x_opt = (14.095, 0.84296)
         f_opt = -6961.814744
@@ -1347,6 +1354,8 @@ class TestDifferentialEvolutionSolver:
 
     def test_L7(self):
         # Lampinen ([5]) test problem 7
+        rng = np.random.default_rng(209331234)
+
         def f(x):
             x = np.hstack(([0], x))  # 1-indexed to match reference
             fun = (5.3578547*x[3]**2 + 0.8356891*x[1]*x[5] +
@@ -1371,7 +1380,7 @@ class TestDifferentialEvolutionSolver:
         bounds = [(78, 102), (33, 45)] + [(27, 45)]*3
         constraints = (N)
 
-        res = differential_evolution(f, bounds, strategy='rand1bin', rng=1234,
+        res = differential_evolution(f, bounds, strategy='rand1bin', rng=rng,
                                      constraints=constraints)
 
         # using our best solution, rather than Lampinen/Koziel. Koziel solution
@@ -1395,6 +1404,8 @@ class TestDifferentialEvolutionSolver:
     @pytest.mark.xfail(platform.machine() == 'ppc64le',
                        reason="fails on ppc64le")
     def test_L8(self):
+        rng = np.random.default_rng(209331234)
+
         def f(x):
             x = np.hstack(([0], x))  # 1-indexed to match reference
             fun = 3*x[1] + 0.000001*x[1]**3 + 2*x[2] + 0.000002/3*x[2]**3
@@ -1428,7 +1439,7 @@ class TestDifferentialEvolutionSolver:
             # huge amount of CPU time. Changing strategy to best1bin speeds
             # things up a lot
             res = differential_evolution(f, bounds, strategy='best1bin',
-                                         rng=1234, constraints=constraints,
+                                         rng=rng, constraints=constraints,
                                          maxiter=5000)
 
         x_opt = (679.9453, 1026.067, 0.1188764, -0.3962336)
@@ -1448,6 +1459,7 @@ class TestDifferentialEvolutionSolver:
     @pytest.mark.fail_slow(5)
     def test_L9(self):
         # Lampinen ([5]) test problem 9
+        rng = np.random.default_rng(209331234)
 
         def f(x):
             x = np.hstack(([0], x))  # 1-indexed to match reference
@@ -1461,7 +1473,7 @@ class TestDifferentialEvolutionSolver:
 
         bounds = [(-1, 1)]*2
         constraints = (N)
-        res = differential_evolution(f, bounds, strategy='rand1bin', rng=1234,
+        res = differential_evolution(f, bounds, strategy='rand1bin', rng=rng,
                                      constraints=constraints)
 
         x_opt = [np.sqrt(2)/2, 0.5]
@@ -1597,10 +1609,12 @@ class TestDifferentialEvolutionSolver:
             return rosen(x)
 
         bounds = [(0, 10), (0, 10)]
+        rng = np.random.default_rng(209331234)
         res1 = differential_evolution(rosen, bounds, updating='deferred',
-                                      rng=1)
+                                      rng=rng)
+        rng = np.random.default_rng(209331234)
         res2 = differential_evolution(rosen_vec, bounds, vectorized=True,
-                                      updating='deferred', rng=1)
+                                      updating='deferred', rng=rng)
 
         # the two minimisation runs should be functionally equivalent
         assert_allclose(res1.x, res2.x)
@@ -1625,11 +1639,13 @@ class TestDifferentialEvolutionSolver:
 
         bounds = [(0, 10), (0, 10)]
 
+        rng = np.random.default_rng(209331234)
         res1 = differential_evolution(rosen, bounds, updating='deferred',
-                                      rng=1, constraints=[nlc1, nlc2],
+                                      rng=rng, constraints=[nlc1, nlc2],
                                       polish=False)
+        rng = np.random.default_rng(209331234)
         res2 = differential_evolution(rosen_vec, bounds, vectorized=True,
-                                      updating='deferred', rng=1,
+                                      updating='deferred', rng=rng,
                                       constraints=[nlc1, nlc2],
                                       polish=False)
         # the two minimisation runs should be functionally equivalent
@@ -1644,8 +1660,14 @@ class TestDifferentialEvolutionSolver:
             return pf.res
         pf.res = None
 
-        res = differential_evolution(rosen, self.bounds, polish=pf, maxiter=1, rng=0)
-        ref = differential_evolution(rosen, self.bounds, polish=True, maxiter=1, rng=0)
+        rng = np.random.default_rng(209331234)
+        res = differential_evolution(
+            rosen, self.bounds, polish=pf, maxiter=1, rng=rng
+        )
+        rng = np.random.default_rng(209331234)
+        ref = differential_evolution(
+            rosen, self.bounds, polish=True, maxiter=1, rng=rng
+        )
 
         # res.success will be False because of the small number of iterations
         # The solution produced by DE would be bad after only one iteration.
@@ -1694,6 +1716,7 @@ class TestDifferentialEvolutionSolver:
         assert_allclose(res.fun, res2.fun)
 
     def test_constraint_violation_error_message(self):
+        rng = np.random.default_rng(209331234)
 
         def func(x):
             return np.cos(x[0]) + np.sin(x[1])
@@ -1707,7 +1730,7 @@ class TestDifferentialEvolutionSolver:
                                         constraints=[c0, c1],
                                         maxiter=10,
                                         polish=False,
-                                        rng=864197532)
+                                        rng=rng)
         assert result.success is False
         # The numerical value in the error message might be sensitive to
         # changes in the implementation.  It can be updated if the code is
@@ -1718,6 +1741,8 @@ class TestDifferentialEvolutionSolver:
 
     @pytest.mark.fail_slow(20)  # fail-slow exception by request - see gh-20806
     def test_strategy_fn(self):
+        rng = np.random.default_rng(209331234)
+
         # examines ability to customize strategy by mimicking one of the
         # in-built strategies
         parameter_count = 4
@@ -1757,7 +1782,7 @@ class TestDifferentialEvolutionSolver:
             mutation=mutation,
             maxiter=2,
             strategy=custom_strategy_fn,
-            rng=10,
+            rng=rng,
             polish=False
         )
         assert solver.strategy is custom_strategy_fn
@@ -1765,8 +1790,9 @@ class TestDifferentialEvolutionSolver:
         assert calls[0] > 0
 
         # check custom strategy works with updating='deferred'
+        rng = np.random.default_rng(209331234)
         res = differential_evolution(
-            rosen, bounds, strategy=custom_strategy_fn, updating='deferred'
+            rosen, bounds, strategy=custom_strategy_fn, updating='deferred', rng=rng
         )
         assert res.success
 

--- a/scipy/optimize/tests/test__differential_evolution.py
+++ b/scipy/optimize/tests/test__differential_evolution.py
@@ -1065,7 +1065,7 @@ class TestDifferentialEvolutionSolver:
         bounds = [(0, 1)]*9 + [(0, 100)]*3 + [(0, 1)]
 
         # using a lower popsize to speed the test up
-        rng = rng = np.random.default_rng(np.random.RandomState(12345))
+        rng = np.random.default_rng(np.random.RandomState(12345))
         res = differential_evolution(
             f, bounds, strategy='best1bin', rng=rng, constraints=(L,),
             popsize=5, tol=0.01
@@ -1089,10 +1089,10 @@ class TestDifferentialEvolutionSolver:
         L = LinearConstraint(csr_array(A), -np.inf, b)
 
         # using a lower popsize to speed the test up
-        rng = np.random.default_rng(np.random.RandomState(1211134))
+        rng = np.random.default_rng(np.random.RandomState(12345))
         res = differential_evolution(
             f, bounds, strategy='best1bin', rng=rng, constraints=(L,),
-            popsize=2, tol=0.05
+            popsize=5, tol=0.01
         )
 
         assert_allclose(f(x_opt), f_opt)
@@ -1122,11 +1122,11 @@ class TestDifferentialEvolutionSolver:
         constraints = (L, N, L2, N2)
 
         with warnings.catch_warnings():
-            rng = np.random.default_rng(np.random.RandomState(1211134))
+            rng = np.random.default_rng(np.random.RandomState(12345))
             warnings.simplefilter("ignore", UserWarning)
             res = differential_evolution(
                 f, bounds, strategy='best1bin', rng=rng,
-                constraints=constraints, popsize=2, tol=0.05
+                constraints=constraints, popsize=5, tol=0.05
             )
 
         assert_allclose(res.x, x_opt, atol=6e-4)

--- a/scipy/optimize/tests/test__differential_evolution.py
+++ b/scipy/optimize/tests/test__differential_evolution.py
@@ -384,7 +384,8 @@ class TestDifferentialEvolutionSolver:
         result = differential_evolution(quadratic,
                                         bounds,
                                         args=args,
-                                        polish=True)
+                                        polish=True,
+                                        rng=1980983098)
         assert_almost_equal(result.fun, 2 / 3.)
 
     def test_init_with_invalid_strategy(self):
@@ -412,7 +413,9 @@ class TestDifferentialEvolutionSolver:
                           bounds)
 
         # test that we can use a new-type Bounds object
-        result = differential_evolution(rosen, Bounds([0, 0], [2, 2]))
+        result = differential_evolution(
+            rosen, Bounds([0, 0], [2, 2]), rng=1980983098
+        )
         assert_almost_equal(result.x, (1., 1.))
 
     def test_select_samples(self):
@@ -437,6 +440,8 @@ class TestDifferentialEvolutionSolver:
     def test_maxfun_stops_solve(self):
         # test that if the maximum number of function evaluations is exceeded
         # during initialisation the solver stops
+        rng = np.random.default_rng(1980983098)
+
         solver = DifferentialEvolutionSolver(rosen, self.bounds, maxfun=1,
                                              polish=False)
         result = solver.solve()
@@ -456,7 +461,8 @@ class TestDifferentialEvolutionSolver:
                                              self.bounds,
                                              popsize=5,
                                              polish=False,
-                                             maxfun=40)
+                                             maxfun=40,
+                                             rng=rng)
         result = solver.solve()
 
         assert_equal(result.nfev, 41)
@@ -473,7 +479,8 @@ class TestDifferentialEvolutionSolver:
                                              popsize=5,
                                              polish=False,
                                              maxfun=47,
-                                             updating='deferred')
+                                             updating='deferred',
+                                             rng=rng)
         result = solver.solve()
 
         assert_equal(result.nfev, 47)
@@ -495,7 +502,8 @@ class TestDifferentialEvolutionSolver:
         differential_evolution(self.quadratic,
                                [(-100, 100)],
                                tol=0.02,
-                               seed=1)
+                               seed=1,
+                               rng=1980983098)
 
     def test_rng_gives_repeatability(self):
         result = differential_evolution(self.quadratic,
@@ -563,8 +571,9 @@ class TestDifferentialEvolutionSolver:
     def test_iteration(self):
         # test that DifferentialEvolutionSolver is iterable
         # if popsize is 3, then the overall generation has size (6,)
+        rng = np.random.default_rng(4210919)
         solver = DifferentialEvolutionSolver(rosen, self.bounds, popsize=3,
-                                             maxfun=12)
+                                             maxfun=12, rng=rng)
         x, fun = next(solver)
         assert_equal(np.size(x, 0), 2)
 
@@ -576,7 +585,7 @@ class TestDifferentialEvolutionSolver:
         assert_raises(StopIteration, next, solver)
 
         # check a proper minimisation can be done by an iterable solver
-        solver = DifferentialEvolutionSolver(rosen, self.bounds)
+        solver = DifferentialEvolutionSolver(rosen, self.bounds, rng=rng)
         _, fun_prev = next(solver)
         for i, soln in enumerate(solver):
             x_current, fun_current = soln
@@ -587,10 +596,11 @@ class TestDifferentialEvolutionSolver:
                 break
 
     def test_convergence(self):
+        rng = np.random.default_rng(4210919)
         solver = DifferentialEvolutionSolver(rosen, self.bounds, tol=0.2,
-                                             polish=False)
+                                             polish=False, rng=rng)
         solver.solve()
-        assert_(solver.convergence < 0.2)
+        assert solver.convergence < 0.2
 
     def test_maxiter_none_GH5731(self):
         # Pre 0.17 the previous default for maxiter and maxfun was None.
@@ -673,7 +683,8 @@ class TestDifferentialEvolutionSolver:
 
     def test_x0(self):
         # smoke test that checks that x0 is usable.
-        res = differential_evolution(rosen, self.bounds, x0=[0.2, 0.8])
+        rng = np.random.default_rng(4210919)
+        res = differential_evolution(rosen, self.bounds, x0=[0.2, 0.8], rng=rng)
         assert res.success
 
         # check what happens if some of the x0 lay outside the bounds
@@ -692,16 +703,19 @@ class TestDifferentialEvolutionSolver:
 
     def test_deferred_updating(self):
         # check setting of deferred updating, with default workers
+        rng = np.random.default_rng(4210919)
         bounds = [(0., 2.), (0., 2.)]
-        solver = DifferentialEvolutionSolver(rosen, bounds, updating='deferred')
-        assert_(solver._updating == 'deferred')
-        assert_(solver._mapwrapper._mapfunc is map)
+        solver = DifferentialEvolutionSolver(
+            rosen, bounds, updating='deferred', rng=rng
+        )
+        assert solver._updating == 'deferred'
+        assert solver._mapwrapper._mapfunc is map
         res = solver.solve()
         assert res.success
 
         # check that deferred updating works with an exponential crossover
         res = differential_evolution(
-            rosen, bounds, updating='deferred', strategy='best1exp'
+            rosen, bounds, updating='deferred', strategy='best1exp', rng=rng
         )
         assert res.success
 
@@ -740,9 +754,10 @@ class TestDifferentialEvolutionSolver:
             solver.solve()
 
     def test_converged(self):
-        solver = DifferentialEvolutionSolver(rosen, [(0, 2), (0, 2)])
+        rng = np.random.default_rng(4210919)
+        solver = DifferentialEvolutionSolver(rosen, [(0, 2), (0, 2)], rng=rng)
         solver.solve()
-        assert_(solver.converged())
+        assert solver.converged()
 
     def test_constraint_violation_fn(self):
         def constr_f(x):
@@ -847,13 +862,15 @@ class TestDifferentialEvolutionSolver:
             assert cv.shape == (2, 3)
 
     def test_constraint_solve(self):
+        rng = np.random.default_rng(4210919)
+
         def constr_f(x):
             return np.array([x[0] + x[1]])
 
         nlc = NonlinearConstraint(constr_f, -np.inf, 1.9)
 
         solver = DifferentialEvolutionSolver(rosen, [(0, 2), (0, 2)],
-                                             constraints=(nlc,))
+                                             constraints=(nlc,), rng=rng)
 
         # trust-constr warns if the constraint function is linear
         with warns(UserWarning):
@@ -864,13 +881,15 @@ class TestDifferentialEvolutionSolver:
 
     @pytest.mark.fail_slow(10)
     def test_impossible_constraint(self):
+        rng = np.random.default_rng(4210919)
+
         def constr_f(x):
             return np.array([x[0] + x[1]])
 
         nlc = NonlinearConstraint(constr_f, -np.inf, -1)
 
         solver = DifferentialEvolutionSolver(
-            rosen, [(0, 2), (0, 2)], constraints=(nlc,), popsize=1, rng=1, maxiter=100
+            rosen, [(0, 2), (0, 2)], constraints=(nlc,), popsize=1, rng=rng, maxiter=100
         )
 
         # a UserWarning is issued because the 'trust-constr' polishing is

--- a/scipy/optimize/tests/test__differential_evolution.py
+++ b/scipy/optimize/tests/test__differential_evolution.py
@@ -1040,7 +1040,6 @@ class TestDifferentialEvolutionSolver:
     @pytest.mark.fail_slow(20)
     def test_L1(self):
         # Lampinen ([5]) test problem 1
-        rng = np.random.default_rng(12345)
 
         def f(x):
             x = np.hstack(([0], x))  # 1-indexed to match reference
@@ -1066,6 +1065,7 @@ class TestDifferentialEvolutionSolver:
         bounds = [(0, 1)]*9 + [(0, 100)]*3 + [(0, 1)]
 
         # using a lower popsize to speed the test up
+        rng = rng = np.random.default_rng(np.random.RandomState(12345))
         res = differential_evolution(
             f, bounds, strategy='best1bin', rng=rng, constraints=(L,),
             popsize=5, tol=0.01
@@ -1089,7 +1089,7 @@ class TestDifferentialEvolutionSolver:
         L = LinearConstraint(csr_array(A), -np.inf, b)
 
         # using a lower popsize to speed the test up
-        rng = np.random.default_rng(12345)
+        rng = np.random.default_rng(np.random.RandomState(1211134))
         res = differential_evolution(
             f, bounds, strategy='best1bin', rng=rng, constraints=(L,),
             popsize=2, tol=0.05
@@ -1122,7 +1122,7 @@ class TestDifferentialEvolutionSolver:
         constraints = (L, N, L2, N2)
 
         with warnings.catch_warnings():
-            rng = np.random.default_rng(12345)
+            rng = np.random.default_rng(np.random.RandomState(1211134))
             warnings.simplefilter("ignore", UserWarning)
             res = differential_evolution(
                 f, bounds, strategy='best1bin', rng=rng,

--- a/scipy/optimize/tests/test__differential_evolution.py
+++ b/scipy/optimize/tests/test__differential_evolution.py
@@ -1229,66 +1229,7 @@ class TestDifferentialEvolutionSolver:
         assert_(np.all(res.x >= np.array(bounds)[:, 0]))
         assert_(np.all(res.x <= np.array(bounds)[:, 1]))
 
-    """
-    # test commented out because of test flakes in macosx-x86_64 that
-    # were also seen in GUIX. It's suspected that they're due to
-    # numeric instability. Given that constraint handling is
-    # adequately tested in the other problems we can skip this one.
-    # It's left in the test suite in case we want to use it again
-    # in the future (and a reminder what the problems with it were).
-    @pytest.mark.fail_slow(10)
-    def test_L4(self):
-        # Lampinen ([5]) test problem 4
-        def f(x):
-            return np.sum(x[:3])
-
-        A = np.zeros((4, 9))
-        A[1, [4, 6]] = 0.0025, 0.0025
-        A[2, [5, 7, 4]] = 0.0025, 0.0025, -0.0025
-        A[3, [8, 5]] = 0.01, -0.01
-        A = A[1:, 1:]
-        b = np.array([1, 1, 1])
-
-        def c1(x):
-            x = np.hstack(([0], x))  # 1-indexed to match reference
-            return [x[1]*x[6] - 833.33252*x[4] - 100*x[1] + 83333.333,
-                    x[2]*x[7] - 1250*x[5] - x[2]*x[4] + 1250*x[4],
-                    x[3]*x[8] - 1250000 - x[3]*x[5] + 2500*x[5]]
-
-        L = LinearConstraint(A, -np.inf, 1)
-        N = NonlinearConstraint(c1, 0, np.inf)
-
-        bounds = [(100, 10000)] + [(1000, 10000)]*2 + [(10, 1000)]*5
-        constraints = (L, N)
-
-        with warnings.catch_warnings():
-            warnings.simplefilter("ignore", UserWarning)
-            res = differential_evolution(
-                f, bounds, strategy='best1bin', rng=1234,
-                constraints=constraints, popsize=3, tol=0.05
-            )
-
-        f_opt = 7049.248
-
-        x_opt = [579.306692, 1359.97063, 5109.9707, 182.0177, 295.601172,
-                217.9823, 286.416528, 395.601172]
-
-        assert_allclose(f(x_opt), f_opt, atol=0.001)
-        assert_allclose(res.fun, f_opt, atol=0.001)
-
-        # use higher tol here for 32-bit Windows, see gh-11693
-        if (platform.system() == 'Windows' and np.dtype(np.intp).itemsize < 8):
-            assert_allclose(res.x, x_opt, rtol=2.4e-6, atol=0.0035)
-        else:
-            # tolerance determined from macOS + MKL failure, see gh-12701
-            assert_allclose(res.x, x_opt, rtol=5e-6, atol=0.0024)
-
-        assert res.success
-        assert_(np.all(A @ res.x <= b))
-        assert_(np.all(np.array(c1(res.x)) >= 0))
-        assert_(np.all(res.x >= np.array(bounds)[:, 0]))
-        assert_(np.all(res.x <= np.array(bounds)[:, 1]))
-    """
+    # test_L4 was remove for flakiness due to numerical instability (gh24549)
 
     @pytest.mark.fail_slow(10)
     def test_L5(self):

--- a/scipy/optimize/tests/test__differential_evolution.py
+++ b/scipy/optimize/tests/test__differential_evolution.py
@@ -1065,7 +1065,7 @@ class TestDifferentialEvolutionSolver:
         bounds = [(0, 1)]*9 + [(0, 100)]*3 + [(0, 1)]
 
         # using a lower popsize to speed the test up
-        rng = np.random.default_rng(np.random.RandomState(12345))
+        rng = np.random.default_rng(np.random.MT19937(12345))
         res = differential_evolution(
             f, bounds, strategy='best1bin', rng=rng, constraints=(L,),
             popsize=5, tol=0.01
@@ -1089,7 +1089,7 @@ class TestDifferentialEvolutionSolver:
         L = LinearConstraint(csr_array(A), -np.inf, b)
 
         # using a lower popsize to speed the test up
-        rng = np.random.default_rng(np.random.RandomState(12345))
+        rng = np.random.default_rng(np.random.MT19937(12345))
         res = differential_evolution(
             f, bounds, strategy='best1bin', rng=rng, constraints=(L,),
             popsize=5, tol=0.01
@@ -1122,7 +1122,7 @@ class TestDifferentialEvolutionSolver:
         constraints = (L, N, L2, N2)
 
         with warnings.catch_warnings():
-            rng = np.random.default_rng(np.random.RandomState(12345))
+            rng = np.random.default_rng(np.random.MT19937(12345))
             warnings.simplefilter("ignore", UserWarning)
             res = differential_evolution(
                 f, bounds, strategy='best1bin', rng=rng,

--- a/scipy/optimize/tests/test__differential_evolution.py
+++ b/scipy/optimize/tests/test__differential_evolution.py
@@ -502,7 +502,6 @@ class TestDifferentialEvolutionSolver:
         differential_evolution(self.quadratic,
                                [(-100, 100)],
                                tol=0.02,
-                               seed=1,
                                rng=1980983098)
 
     def test_rng_gives_repeatability(self):

